File: C#/OptimizationMethods/MultiDimensional.cs
using MathUtils;
using System;

namespace OptimizationMethods
{
    public static class MultiDimensional
    {
        ////////////////////
        /// Lab. work #2 ///
        ////////////////////
        public static DoubleVector BiSect(FunctionND targetFunction, DoubleVector lhs, DoubleVector rhs, double accuracy, int iterations)
        {
            DoubleVector x_c, dir;
            dir = DoubleVector.Direction(lhs, rhs) * accuracy;
            int cntr = 0;
            for (; cntr != iterations && DoubleVector.Distance(rhs, lhs) > 2 * accuracy; cntr++)
            {
                x_c = (rhs + lhs) * 0.5;
                if (targetFunction(x_c + dir) > targetFunction(x_c - dir))
                    rhs = x_c;
                else
                    lhs = x_c;
            }
#if DEBUG
            Console.WriteLine($"BiSect iterations number : {cntr}");
            Console.WriteLine($"BiSect argument range    : {DoubleVector.Distance(rhs, lhs)}");
#endif
            return (rhs + lhs) * 0.5;
        }
        public static DoubleVector BiSect(FunctionND targetFunction, DoubleVector lhs, DoubleVector rhs, double accuracy) => BiSect(targetFunction, lhs, rhs, accuracy, NumericCommon.ITERATIONS_COUNT_HIGH);
        public static DoubleVector BiSect(FunctionND targetFunction, DoubleVector lhs, DoubleVector rhs) => BiSect(targetFunction, lhs, rhs, NumericCommon.NUMERIC_ACCURACY_MIDDLE, NumericCommon.ITERATIONS_COUNT_HIGH);
        public static DoubleVector GoldenRatio(FunctionND targetFunction, DoubleVector _lhs, DoubleVector _rhs, double accuracy, int iterations)
        {
            DoubleVector lhs = new DoubleVector(_lhs);
            DoubleVector rhs = new DoubleVector(_rhs);
            DoubleVector x_l = rhs - (rhs - lhs) * NumericCommon.PSI;
            DoubleVector x_r = lhs + (rhs - lhs) * NumericCommon.PSI;
            double f_l = targetFunction(x_l);
            double f_r = targetFunction(x_r);
            int iteration = 0;
            for (; iteration != iterations && DoubleVector.Distance(rhs, lhs) > 2 * accuracy; iteration++)
            {
                if (f_l > f_r)
                {
                    lhs = x_l;
                    x_l = x_r;
                    f_l = f_r;
                    x_r = lhs + (rhs - lhs) * NumericCommon.PSI;
                    f_r = targetFunction(x_r);
                }
                else
                {
                    rhs = x_r;
                    x_r = x_l;
                    f_r = f_l;
                    x_l = rhs - (rhs - lhs) * NumericCommon.PSI;
                    f_l = targetFunction(x_l);
                }
            }
#if DEBUG
            Console.WriteLine($"GoldenRatio iterations number : {iteration + 2}");
            Console.WriteLine($"GoldenRatio argument range    : {DoubleVector.Distance(rhs, lhs)}");
#endif
            return (rhs + lhs) * 0.5;
        }
        public static DoubleVector GoldenRatio(FunctionND targetFunction, DoubleVector lhs, DoubleVector rhs, double accuracy) => GoldenRatio(targetFunction, lhs, rhs, accuracy, NumericCommon.ITERATIONS_COUNT_HIGH);
        public static DoubleVector GoldenRatio(FunctionND targetFunction, DoubleVector lhs, DoubleVector rhs) => GoldenRatio(targetFunction, lhs, rhs, NumericCommon.NUMERIC_ACCURACY_MIDDLE, NumericCommon.ITERATIONS_COUNT_HIGH);
        public static DoubleVector Fibonacci(FunctionND targetFunction, DoubleVector _lhs, DoubleVector _rhs, double accuracy)
        {
            DoubleVector lhs = new DoubleVector(_lhs);
            DoubleVector rhs = new DoubleVector(_rhs);
            double fib_t = 0.0, fib_1 = 1.0, fib_2 = 1.0;
            double condition = DoubleVector.Distance(rhs, lhs) / accuracy;
            int iterations = 0;
            while (fib_2 < condition)
            {
                iterations++;
                fib_t = fib_1;
                fib_1 = fib_2;
                fib_2 += fib_t;
            }
            DoubleVector x_l = lhs + (rhs - lhs) * ((fib_2 - fib_1) / fib_2);
            DoubleVector x_r = lhs + (rhs - lhs) * (fib_1 / fib_2);

            double f_l = targetFunction(x_l);
            double f_r = targetFunction(x_r);
            for (int index = iterations; index > 0; index--)
            {
                fib_t = fib_2 - fib_1;
                fib_2 = fib_1;
                fib_1 = fib_t;

                if (f_l > f_r)
                {
                    lhs = x_l;
                    f_l = f_r;
                    x_l = x_r;
                    x_r = lhs + (rhs - lhs) * (fib_1 / fib_2);
                    f_r = targetFunction(x_r);
                }
                else
                {
                    rhs = x_r;
                    x_r = x_l;
                    f_r = f_l;
                    x_l = lhs + (rhs - lhs) * ((fib_2 - fib_1) / fib_2);
                    f_l = targetFunction(x_l);
                }
            }
#if DEBUG
            Console.WriteLine($"Fibonacci iterations number : {iterations + 2}");
            Console.WriteLine($"Fibonacci argument range    : {DoubleVector.Distance(rhs, lhs)}");
#endif
            return (rhs + lhs) * 0.5;
        }
        public static DoubleVector Fibonacci(FunctionND targetFunction, DoubleVector lhs, DoubleVector rhs) => Fibonacci(targetFunction, lhs, rhs, NumericCommon.NUMERIC_ACCURACY_MIDDLE);

        public static DoubleVector PerCoordDescend(FunctionND targetFunction, DoubleVector xStart, double accuracy, int iterations)
        {
            DoubleVector lhs = new DoubleVector(xStart);

            DoubleVector rhs = new DoubleVector(xStart);

            double step = 1.0;

            double x_i, y_1, y_0;

            int opt_coord_n = 0, coord_id;
            
            int i = 0;

            for (i = 0; i < iterations; i++)
            {
                coord_id = i % lhs.Count;

                rhs[coord_id] -= accuracy;

                y_0 = targetFunction(rhs);

                rhs[coord_id] += 2 * accuracy;

                y_1 = targetFunction(rhs);

                rhs[coord_id] -= accuracy;

                rhs[coord_id] = y_0 > y_1 ? rhs[coord_id] += step : rhs[coord_id] -= step;

                x_i = lhs[coord_id];

                rhs = BiSect(targetFunction, lhs, rhs, accuracy, iterations);

                lhs = new DoubleVector(rhs);

                if (Math.Abs(rhs[coord_id] - x_i) < accuracy)
                {
                    opt_coord_n++;

                    if (opt_coord_n == rhs.Count)
                    {
#if DEBUG
                        Console.WriteLine($"per coord descend iterations number : {i}");
#endif
                        return lhs;
                    }
                    continue;
                }
                opt_coord_n = 0;
            }
#if DEBUG
            Console.WriteLine($"per coord descend iterations number : {iterations}");
#endif
            return lhs;
        }
        public static DoubleVector PerCoordDescend(FunctionND targetFunction, DoubleVector xStart, double accuracy) => PerCoordDescend(targetFunction, xStart, accuracy, NumericCommon.ITERATIONS_COUNT_HIGH);
        public static DoubleVector PerCoordDescend(FunctionND targetFunction, DoubleVector xStart) => PerCoordDescend(targetFunction, xStart, NumericCommon.NUMERIC_ACCURACY_MIDDLE, NumericCommon.ITERATIONS_COUNT_HIGH);

        ////////////////////
        /// Lab. work #3 ///
        ////////////////////
        public static DoubleVector GradientDescend(FunctionND targetFunction, DoubleVector xStart, double accuracy, int iterations)
        {
            DoubleVector x_i = new DoubleVector(xStart);

            DoubleVector x_i_1 = new DoubleVector(xStart);

            int cntr = 0;

            for (; cntr <= iterations; cntr++)
            {
                x_i_1 = x_i - DoubleVector.Gradient(targetFunction, x_i, accuracy);

                x_i_1 = BiSect(targetFunction, x_i, x_i_1, accuracy, iterations);

                if ((x_i_1 - x_i).Magnitude < accuracy) break;

                x_i = x_i_1;
            }
#if DEBUG
            Console.WriteLine($"gradient descend iterations number : {cntr}");
#endif
            return (x_i_1 + x_i) * 0.5;
        }
        public static DoubleVector GradientDescend(FunctionND targetFunction, DoubleVector xStart, double accuracy) => GradientDescend(targetFunction, xStart, accuracy, NumericCommon.ITERATIONS_COUNT_HIGH);
        public static DoubleVector GradientDescend(FunctionND targetFunction, DoubleVector xStart) => GradientDescend(targetFunction, xStart, NumericCommon.NUMERIC_ACCURACY_MIDDLE, NumericCommon.ITERATIONS_COUNT_HIGH);
        
        public static DoubleVector СonjGradientDescend(FunctionND targetFunction, DoubleVector xStart, double accuracy, int iterations)
        {
            DoubleVector x_i = new DoubleVector(xStart);

            DoubleVector x_i_1 = new DoubleVector(xStart);

            DoubleVector s_i = DoubleVector.Gradient(targetFunction, xStart, accuracy) * (-1.0), s_i_1;

            double omega;

            int cntr = 0;

            for (; cntr <= iterations; cntr++)
            {
                x_i_1 = x_i + s_i;

                x_i_1 = BiSect(targetFunction, x_i, x_i_1, accuracy, iterations);

                if ((x_i_1 - x_i).Magnitude < accuracy) break;

                s_i_1 = DoubleVector.Gradient(targetFunction, x_i_1, accuracy);

                omega = Math.Pow((s_i_1).Magnitude, 2) / Math.Pow((s_i).Magnitude, 2);

                s_i = s_i * omega - s_i_1;

                x_i = x_i_1;
            }
#if DEBUG
            Console.WriteLine($"conj gradient descend iterations number : {cntr}");
#endif
            return (x_i_1 + x_i) * 0.5;
        }
        public static DoubleVector СonjGradientDescend(FunctionND targetFunction, DoubleVector xStart, double accuracy) => СonjGradientDescend(targetFunction, xStart, accuracy, NumericCommon.ITERATIONS_COUNT_HIGH);
        public static DoubleVector СonjGradientDescend(FunctionND targetFunction, DoubleVector xStart) => СonjGradientDescend(targetFunction, xStart, NumericCommon.NUMERIC_ACCURACY_MIDDLE, NumericCommon.ITERATIONS_COUNT_HIGH);
    
        ////////////////////
        /// Lab. work #4 ///
        ////////////////////
        public static DoubleVector NewtoneRaphson(FunctionND targetFunction, DoubleVector xStart, double accuracy, int iterations)
        {
            DoubleVector x_i   = new DoubleVector(xStart);

            DoubleVector x_i_1 = new DoubleVector(xStart);

            int cntr = 0;

            for (; cntr <= iterations; cntr++)
            { 
                x_i_1 = x_i - DoubleMatrix.Invert(DoubleMatrix.Hessian(targetFunction, x_i, accuracy)) * DoubleVector.Gradient(targetFunction, x_i, accuracy);

                if ((x_i_1 - x_i).Magnitude < accuracy) break;

                x_i = x_i_1;
            }
#if DEBUG
            Console.WriteLine($"Newtone - Raphson iterations number : {cntr}");
#endif
            return (x_i_1 + x_i) * 0.5;
        }
        public static DoubleVector NewtoneRaphson(FunctionND targetFunction, DoubleVector xStart, double accuracy) => NewtoneRaphson(targetFunction, xStart, accuracy, NumericCommon.ITERATIONS_COUNT_HIGH);
        public static DoubleVector NewtoneRaphson(FunctionND targetFunction, DoubleVector xStart) => NewtoneRaphson(targetFunction, xStart, NumericCommon.NUMERIC_ACCURACY_MIDDLE, NumericCommon.ITERATIONS_COUNT_HIGH);

    }
}

File: Java/OptimizationMethods/src/MultiDimensional.java
import functionalInterfaces.IFunctionND;
import mathUtils.NumericCommon;
import mathUtils.DoubleVector;
import mathUtils.DoubleMatrix;
import mathUtils.NumericUtils;

import java.util.Objects;

@SuppressWarnings("all")
public class MultiDimensional {
    ////////////////////
    /// Lab. work #2 ///
    ////////////////////
    public static DoubleVector biSect(IFunctionND f, DoubleVector x0, DoubleVector x1,
                                      double eps, int maxIterations) {
        DoubleVector x_0 = new DoubleVector(x0), x_1 = new DoubleVector(x1), x_c, dir;
        dir = DoubleVector.direction(x_0, x_1).mul(eps);
        int iteration = 0;
        for (; iteration != maxIterations; iteration++) {
            if (DoubleVector.sub(x_1, x_0).magnitude() < eps) break;
            x_c = DoubleVector.add(x_1, x_0).mul(0.5);
            if (f.call(DoubleVector.add(x_c, dir)) > f.call(DoubleVector.sub(x_c, dir))) {
                x_1 = x_c;
                continue;
            }
            x_0 = x_c;
        }
        if (NumericCommon.SHOW_ZERO_ORDER_METHODS_DEBUG_LOG) System.out.printf("BiSect iterations number : %s\n", iteration);
        return DoubleVector.add(x_1, x_0).mul(0.5);
    }

    public static DoubleVector biSect(IFunctionND f, DoubleVector x1, DoubleVector x2, double eps) {
        return biSect(f, x1, x2, eps, NumericCommon.ITERATIONS_COUNT_HIGH);
    }

    public static DoubleVector biSect(IFunctionND f, DoubleVector x1, DoubleVector x2) {
        return biSect(f, x1, x2, NumericCommon.NUMERIC_ACCURACY_MIDDLE, NumericCommon.ITERATIONS_COUNT_HIGH);
    }

    public static DoubleVector goldenRatio(IFunctionND f, DoubleVector x1, DoubleVector x2,
                                           double eps, int maxIterations) {
        DoubleVector a = new DoubleVector(x1);
        DoubleVector b = new DoubleVector(x2);
        DoubleVector dx;
        int iteration = 0;
        for (; iteration != maxIterations; iteration++) {
            if (DoubleVector.sub(x2, x1).magnitude() < eps) break;
            dx = DoubleVector.sub(b, a).mul(NumericCommon.ONE_OVER_PHI);
            x1 = DoubleVector.sub(b, dx);
            x2 = DoubleVector.add(a, dx);
            if (f.call(x1) >= f.call(x2)) {
                a = x1;
                continue;
            }
            b = x2;
        }

        if (NumericCommon.SHOW_ZERO_ORDER_METHODS_DEBUG_LOG)
            System.out.printf("golden ratio iterations number : %s\n", iteration);

        return DoubleVector.add(x2, x1).mul(0.5);
    }

    public static DoubleVector goldenRatio(IFunctionND f, DoubleVector x1, DoubleVector x2, double eps) {
        return goldenRatio(f, x1, x2, eps, NumericCommon.ITERATIONS_COUNT_HIGH);
    }

    public static DoubleVector goldenRatio(IFunctionND f, DoubleVector x1, DoubleVector x2) {
        return goldenRatio(f, x1, x2, NumericCommon.NUMERIC_ACCURACY_MIDDLE, NumericCommon.ITERATIONS_COUNT_HIGH);
    }

    public static DoubleVector fibonacci(IFunctionND f, DoubleVector x0, DoubleVector x1, double eps) {
        DoubleVector a = new DoubleVector(x0);
        DoubleVector b = new DoubleVector(x1);
        DoubleVector dx;
        int[] fib_pair = NumericUtils.closestFibonacciPair(DoubleVector.sub(b, a).magnitude() / eps);
        int f_n = fib_pair[0], f_n_1 = fib_pair[1], f_tmp, iteration = 0;
        while (f_n != f_n_1) {
            if (DoubleVector.sub(x1, x0).magnitude() < eps) break;
            iteration++;
            f_tmp = f_n_1 - f_n;
            dx = DoubleVector.sub(b, a);
            x0 = DoubleVector.add(a, DoubleVector.mul(dx, (double) f_tmp / f_n_1));
            x1 = DoubleVector.add(a, DoubleVector.mul(dx, (double) f_n / f_n_1));
            f_n_1 = f_n;
            f_n = f_tmp;
            if (f.call(x0) < f.call(x1)) {
                b = x1;
                continue;
            }
            a = x0;
        }
        if (NumericCommon.SHOW_ZERO_ORDER_METHODS_DEBUG_LOG)
            System.out.printf("fibonacci iterations number : %s\n", iteration);
        return DoubleVector.add(x1, x0).mul(0.5);
    }

    public static DoubleVector fibonacci(IFunctionND f, DoubleVector x0, DoubleVector x1) {
        return fibonacci(f, x0, x1, NumericCommon.NUMERIC_ACCURACY_MIDDLE);
    }

    public static DoubleVector perCordDescend(IFunctionND f, DoubleVector xStart, double eps, int maxIterations) {
        DoubleVector x_0 = new DoubleVector(xStart);
        DoubleVector x_1 = new DoubleVector(xStart);
        double step = 1.0;
        double x_i, y_1, y_0;
        int optCoordinatesCount = 0, coordinateId;
        for (int iteration = 0; iteration < maxIterations; iteration++) {
            coordinateId = iteration % x_0.size();
            x_1.set(coordinateId, x_1.get(coordinateId) - eps);
            y_0 = f.call(x_1);
            x_1.set(coordinateId, x_1.get(coordinateId) + 2 * eps);
            y_1 = f.call(x_1);
            x_1.set(coordinateId, x_1.get(coordinateId) - eps);
            x_1.set(coordinateId, y_0 > y_1 ? x_1.get(coordinateId) + step : x_1.get(coordinateId) - step);
            x_i = x_0.get(coordinateId);
            x_1 = biSect(f, x_0, x_1, eps, maxIterations);
            x_0 = new DoubleVector(x_1);
            if (Math.abs(x_1.get(coordinateId) - x_i) < eps) {
                optCoordinatesCount++;
                if (optCoordinatesCount == x_1.size()) {
                    if (NumericCommon.SHOW_DEBUG_LOG)
                        System.out.printf("per cord descend iterations number : %s\n", iteration + 1);
                    return x_0;
                }
                continue;
            }
            optCoordinatesCount = 0;
        }
        if (NumericCommon.SHOW_DEBUG_LOG) System.out.printf("per cord descend iterations number : %s\n", maxIterations);
        return x_0;
    }

    public static DoubleVector perCordDescend(IFunctionND f, DoubleVector xStart, double eps) {
        return perCordDescend(f, xStart, eps, NumericCommon.ITERATIONS_COUNT_HIGH);
    }

    public static DoubleVector perCordDescend(IFunctionND f, DoubleVector xStart) {
        return perCordDescend(f, xStart, NumericCommon.NUMERIC_ACCURACY_MIDDLE, NumericCommon.ITERATIONS_COUNT_HIGH);
    }

    ////////////////////
    /// Lab. work #3 ///
    ////////////////////
    public static DoubleVector gradientDescend(IFunctionND f, DoubleVector xStart, double eps, int maxIterations) {
        DoubleVector x_i = new DoubleVector(xStart);
        DoubleVector x_i_1 = new DoubleVector(xStart);
        int cntr = 0;
        for (; cntr != maxIterations; cntr++) {
            x_i_1 = DoubleVector.sub(x_i, DoubleVector.gradient(f, x_i, eps));
            x_i_1 = biSect(f, x_i, x_i_1, eps, maxIterations);
            if (DoubleVector.sub(x_i_1, x_i).magnitude() < eps) break;
            x_i = x_i_1;
        }

        if (NumericCommon.SHOW_DEBUG_LOG) System.out.printf("gradient descend iterations number : %s\n", cntr + 1);

        return DoubleVector.add(x_i_1, x_i).mul(0.5);
    }

    public static DoubleVector gradientDescend(IFunctionND f, DoubleVector xStart, double eps) {
        return gradientDescend(f, xStart, eps, NumericCommon.ITERATIONS_COUNT_HIGH);
    }

    public static DoubleVector gradientDescend(IFunctionND f, DoubleVector xStart) {
        return gradientDescend(f, xStart, NumericCommon.NUMERIC_ACCURACY_MIDDLE, NumericCommon.ITERATIONS_COUNT_HIGH);
    }

    public static DoubleVector conjGradientDescend(IFunctionND f, DoubleVector xStart, double eps, int maxIterations) {
        DoubleVector x_i = new DoubleVector(xStart);
        DoubleVector x_i_1 = new DoubleVector(xStart);
        DoubleVector s_i = DoubleVector.gradient(f, xStart, eps).mul(-1.0), s_i_1;
        double omega;
        int iteration = 0;
        for (; iteration != maxIterations; iteration++) {
            x_i_1 = DoubleVector.add(x_i, s_i);
            x_i_1 = biSect(f, x_i, x_i_1, eps, maxIterations);
            if (DoubleVector.sub(x_i_1, x_i).magnitude() < eps) break;
            s_i_1 = DoubleVector.gradient(f, x_i_1, eps);
            omega = Math.pow((s_i_1).magnitude(), 2) / Math.pow((s_i).magnitude(), 2);
            s_i.mul(omega).sub(s_i_1);
            x_i = x_i_1;
        }

        if (NumericCommon.SHOW_DEBUG_LOG) System.out.printf("Conj gradient descend iterations number : %s\n", iteration + 1);
        return DoubleVector.add(x_i_1, x_i).mul(0.5);
    }

    public static DoubleVector conjGradientDescend(IFunctionND f, DoubleVector xStart, double eps) {
        return conjGradientDescend(f, xStart, eps, NumericCommon.ITERATIONS_COUNT_HIGH);
    }

    public static DoubleVector conjGradientDescend(IFunctionND f, DoubleVector xStart) {
        return conjGradientDescend(f, xStart, NumericCommon.NUMERIC_ACCURACY_MIDDLE, NumericCommon.ITERATIONS_COUNT_HIGH);
    }

    ////////////////////
    /// Lab. work #4 ///
    ////////////////////
    public static DoubleVector newtoneRaphson(IFunctionND f, DoubleVector xStart, double eps, int maxIterations) {
        DoubleVector x_i = new DoubleVector(xStart);
        DoubleVector x_i_1 = new DoubleVector(xStart);
        int iteration = -1;
        for (; iteration != maxIterations; iteration++) {
            DoubleMatrix invHessian = Objects.requireNonNull(DoubleMatrix.invert(DoubleMatrix.hessian(f, x_i, eps)));
            x_i_1 = DoubleVector.sub(x_i, DoubleMatrix.mul(invHessian, DoubleVector.gradient(f, x_i, eps)));
            ///Метод работает, но условие снизу не отрабатывает
            if (DoubleVector.sub(x_i_1, x_i).magnitude() < eps) break;
            x_i = x_i_1;
        }

        if (NumericCommon.SHOW_DEBUG_LOG) System.out.printf("Newtone - Raphson iterations number : %s\n", iteration + 1);
        return DoubleVector.add(x_i_1, x_i).mul(0.5);
    }

    public static DoubleVector newtoneRaphson(IFunctionND f, DoubleVector xStart, double eps) {
        return newtoneRaphson(f, xStart, eps, NumericCommon.ITERATIONS_COUNT_LOW);
    }

    public static DoubleVector newtoneRaphson(IFunctionND f, DoubleVector xStart) {
        return newtoneRaphson(f, xStart, NumericCommon.NUMERIC_ACCURACY_MIDDLE, NumericCommon.ITERATIONS_COUNT_LOW);
    }
}
